src/backend/distributed/connection/locally_reserved_shared_connection_stats.c

```diff
@@ -0,0 +1,508 @@
+					64, &reservedCounterInfo, hashFlags);
+}
+
+
+/*
+ *  CanUseReservedConnection returns true if we have already reserved at least
+ *  one shared connection in this session that is not used.
+ */
+bool
+CanUseReservedConnection(const char *hostName, int nodePort, char *userName,
+						 Oid databaseOid)
+{
+	ReservedConnectionCounterHashKey key;
+
+	strlcpy(key.hostname, hostName, MAX_NODE_LENGTH);
+	strlcpy(key.user, userName, NAMEDATALEN);
+	key.port = nodePort;
+	key.databaseOid = databaseOid;
+
+	bool found = false;
+	ReservedConnectionCounterHashEntry *entry =
+		(ReservedConnectionCounterHashEntry *)
+		hash_search(SessionLocalReservedConnectionCounters, &key, HASH_FIND, &found);
+
+	if (!found || !entry)
+	{
+		return false;
+	}
+
+	return !entry->usedReservation;
+}
+
+
+/*
+ * DeallocateReservedConnections is responsible for two things. First, if the operation
+ * has reserved a connection but not used, it gives back the connection back to the
+ * shared memory pool. Second, for all cases, it deallocates the session local entry from
+ * the hash.
+ */
+void
+DeallocateReservedConnections(void)
+{
+	HASH_SEQ_STATUS status;
+	ReservedConnectionCounterHashEntry *entry;
+
+	hash_seq_init(&status, SessionLocalReservedConnectionCounters);
+	while ((entry = (ReservedConnectionCounterHashEntry *) hash_seq_search(&status)) != 0)
+	{
+		if (!entry->usedReservation)
+		{
+			/*
+			 * We have not used this reservation, make sure to clean-up from
+			 * the shared memory as well.
+			 */
+			DecrementSharedConnectionCounter(entry->key.hostname, entry->key.port);
+
+			/* for completeness, set it to true */
+			entry->usedReservation = true;
+		}
+
+		/*
+		 * We cleaned up all the entries because we may not need reserved connections
+		 * in the next iteration.
+		 */
+		bool found = false;
+		hash_search(SessionLocalReservedConnectionCounters, entry, HASH_REMOVE, &found);
+		Assert(found);
+	}
+}
+
+
+/*
+ * MarkReservedConnectionUsed sets the local hash that the reservation is used.
+ */
+void
+MarkReservedConnectionUsed(const char *hostName, int nodePort, char *userName,
+						   Oid databaseOid)
+{
+	ReservedConnectionCounterHashKey key;
+
+	strlcpy(key.hostname, hostName, MAX_NODE_LENGTH);
+	strlcpy(key.user, userName, NAMEDATALEN);
+	key.port = nodePort;
+	key.databaseOid = databaseOid;
+
+	bool found = false;
+	ReservedConnectionCounterHashEntry *entry =
+		(ReservedConnectionCounterHashEntry *) hash_search(
+			SessionLocalReservedConnectionCounters, &key, HASH_FIND, &found);
+
+	if (!found)
+	{
+		ereport(ERROR, (errmsg("BUG: untracked reserved connection"),
+						errhint("Set citus.max_shared_pool_size TO -1 to "
+								"disable reserved connection counters")));
+	}
+
+	/* a reservation can only be used once */
+	Assert(!entry->usedReservation);
+
+	entry->usedReservation = true;
+}
+
+
+/*
+ * EnsurePrimaryNodesHaveReservedConnection is a wrapper around
+ * ReserveSharedConnectionCounterForNodeListIfNeeded.
```

Comment:
Could we align these names somehow?

e.g. EnsureConnectionPossibilityForPrimaryNodes / EnsureConnectionPossibilityForNodeList